File: JavaScript/Linked List/2130. Maximum Twin Sum of a Linked List.js
/* 2130. Maximum Twin Sum of a Linked List
Difficulty: Medium
Submission Link: https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/submissions/1391188024/
*/

var pairSum = function(head) {
    let maxSum = 0;
    let fast = head;
    let slow = head;
    while(fast && fast.next){
        fast = fast.next.next;
        slow = slow.next;
    }
    fast = head;
    slow = reverse(slow);
    
    function reverse(head){
        let prev = null;
        let current = head;
        while(current){
            let next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }
    while(slow){
        let sum = fast.val + slow.val;
        maxSum = Math.max(sum,maxSum);
        slow = slow.next;
        fast = fast.next;
    }
    return maxSum;
};

File: C/Linked List/2130. Maximum Twin Sum of a Linked List.c
/* 2130. Maximum Twin Sum of a Linked List
Difficulty: Medium
Submission Link: https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/submissions/1391189055/ */

int pairSum(struct ListNode* head) {
    int size = 0;
    int maxTwinSum;
    struct ListNode* ptr = head;
    
    while (ptr) {
        size++;
        ptr = ptr->next;
    }

    int* arr = (int*)malloc(size * sizeof(int));
    
    int i = 0, j = size - 1;
    ptr = head;
    while (ptr) {
        arr[i++] = ptr->val;
        ptr = ptr->next;
    }

    maxTwinSum = arr[0] + arr[j];
    for (i = 0; i < size/2; i++) {
        if (arr[i] + arr[j] > maxTwinSum) {
            maxTwinSum = arr[i] + arr[j];
        }
        j--;
    }

    return maxTwinSum;
}